casper/src/test/scala/coop/rchain/casper/batch2/BlockReceiverEffectsSpec.scala

```diff
@@ -0,0 +1,241 @@
+        a2InOutQueue shouldBe a2.blockHash
+      }
+  }
+
+  private def blockDagStorageMock[F[_]: Applicative]: BlockDagStorage[F] = {
+    val emptyDag = DagRepresentation(Set(), Map(), SortedMap(), DagMessageState(), Map())
+    mock[BlockDagStorage[F]].getRepresentation returnsF emptyDag
+  }
+
+  private def blockRetrieverMock[F[_]: Applicative]: BlockRetriever[F] = {
+    val brMock = mock[BlockRetriever[F]]
+    brMock.ackReceived(*) returns ().pure[F]
+    brMock.admitHash(*, *, *) returnsF AdmitHashResult(
+      Ignore,
+      broadcastRequest = false,
+      requestBlock = false
+    )
+    brMock
+  }
+
+  private def blockStoreMock[F[_]: Sync]: BlockStore[F] = {
+    val state  = Ref.unsafe[F, Map[BlockHash, BlockMessage]](Map())
+    val bsMock = mock[BlockStore[F]]
+    bsMock.contains(*) answers { keys: Seq[BlockHash] =>
+      state.get.map(s => Seq(s.contains(keys.head)))
+    }
+    bsMock.put(*) answers { kvPairs: Seq[(BlockHash, BlockMessage)] =>
+      state.update(s => kvPairs.foldLeft(s) { case (acc, item) => acc + item })
+    }
+    bsMock
+  }
+
+  import fs2._
+
+  private def withEnv[F[_]: Concurrent: Log](shardId: String)(
+      f: (
+          Queue[F, BlockMessage],
+          Queue[F, BlockMessage],
+          Stream[F, BlockHash],
+          BlockStore[F],
+          BlockRetriever[F],
+          BlockDagStorage[F]
+      ) => F[Assertion]
+  ): F[Assertion] =
+    for {
+      state                 <- Ref[F].of(BlockReceiverState[BlockHash])
+      incomingBlockQueue    <- Queue.unbounded[F, BlockMessage]
+      incomingBlockStream   = incomingBlockQueue.dequeue
+      validatedBlocksQueue  <- Queue.unbounded[F, BlockMessage]
+      validatedBlocksStream = validatedBlocksQueue.dequeue
+
+      // Create mock separately for each test
+      bs  = blockStoreMock[F]
+      br  = blockRetrieverMock[F]
+      bds = blockDagStorageMock[F]
+
+      blockReceiver <- {
+        implicit val (bsImp, brImp, bdsImp) = (bs, br, bds)
+        BlockReceiver(state, incomingBlockStream, validatedBlocksStream, shardId)
+      }
+      res <- f(incomingBlockQueue, validatedBlocksQueue, blockReceiver, bs, br, bds)
+    } yield res
+
+  private def makeDefaultBlock =
+    BlockMessage
+      .from(
+        BlockMessageProto(
+          shardId = "root",
+          postStateHash = "abc".unsafeHexToByteString,
+          sigAlgorithm = Secp256k1.name
+        )
+      )
+      .right
+      .get
+
+  private def makeBlock(justifications: List[BlockHash] = List()): BlockMessage = {
+    val (privateKey, pubKey) = Secp256k1.newKeyPair
+    val block =
+      makeDefaultBlock.copy(sender = pubKey.bytes.toByteString, justifications = justifications)
+    ValidatorIdentity(privateKey).signBlock(block)
+  }
+
+  private def addBlock[F[_]: Sync](bs: BlockStore[F]): F[BlockMessage] =
+    for {
+      block <- Sync[F].delay(makeBlock())
+      _     <- bs.put(Seq((block.blockHash, block)))
+    } yield block
```

Comment:
just a comment - this can be done shorter
 ```scala 
 private def addBlock[F[_]: Sync](bs: BlockStore[F]): F[BlockMessage] = {
    val block = makeBlock()
    bs.put(Seq((block.blockHash, block))).as(block)
  }
```

Reply:
@nzpr Is this function now _referentially transparent_?

Let's imagine that `makeBlock()` is heavy operation and possibly do some logging to make this example more extreme.

What would be executed when this assignment is created in the old and new version of `addBlock` function?
```scala
val makeBlockInStore = addBlock(blockStore)
```